src/runtime/local/kernels/EigenCal.h

```diff
@@ -0,0 +1,157 @@
+// The eigenvectors are normalized to have (Euclidean) norm equal to one. 
+// ****************************************************************************
+
+template<class DTRes1, class DTRes2, class VTArg>
+struct EigenCal {
+    static void apply(DTRes1 *& res1, DTRes2 *& res2, const VTArg * inMat, DCTX(ctx)) = delete;
+};
+
+// ****************************************************************************
+// Convenience function
+// ****************************************************************************
+
+
+template<class DTRes1, class DTRes2, class VTArg>
+void eigenCal(DTRes1 *& res1, DTRes2 *& res2, const VTArg * inMat, DCTX(ctx)) {
+    EigenCal<DTRes1, DTRes2, VTArg>::apply(res1,res2, inMat, ctx);
+}
+
+// ****************************************************************************
+// (Partial) template specializations for different data/value types
+// ****************************************************************************
+
+// ----------------------------------------------------------------------------
+// DenseMatrix 
+// Double Value types as input
+// ----------------------------------------------------------------------------
+
+template<>
+struct EigenCal<DenseMatrix<double>,DenseMatrix<double>,DenseMatrix<double>> {
+    static void apply(DenseMatrix<double> *& res1,DenseMatrix<double> *& res2, const DenseMatrix<double> * inMat, DCTX(ctx)) {
+        const auto nr = static_cast<size_t>(inMat->getNumRows());
+        const auto nc = static_cast<size_t>(inMat->getNumCols());
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> inputMatrix(nr, nc);
+       
+       std::tuple<int, int>{1, -1};
+        for(size_t r=0; r<nr; r++){
+            for(size_t c=0; c<nc; c++){
+                inputMatrix(r,c)=inMat->get(r,c);
+            }
+        }
+  Eigen::EigenSolver<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> > s(inputMatrix);  // the instance s(A) includes the eigensystem
+ 
+  size_t eigenValuesrows=s.eigenvalues().rows();
+  size_t eigenValuescols=s.eigenvalues().cols();
+
+  size_t eigenVectorsrows=s.eigenvectors().rows();
+  size_t eigenVectorscols=s.eigenvectors().cols();
+  Eigen::MatrixXd eigenVectors = s.eigenvectors().real().cast <double> (); 
+
+    if(res1 == nullptr)
+             res1= DataObjectFactory::create<DenseMatrix<double>>(eigenValuesrows, eigenValuescols, false);
+
+    if(res2 == nullptr)
+             res2= DataObjectFactory::create<DenseMatrix<double>>(eigenVectorsrows, eigenVectorscols, false);
+
+    for(size_t r = 0; r < eigenValuesrows; r++) {
+                for(size_t c = 0; c < eigenValuescols; c++) {             
+                    res1->set(r,c,s.eigenvalues()[r].real());
+    }
+    }
+    for(size_t r = 0; r < eigenVectorsrows; r++) {
+                for(size_t c = 0; c < eigenVectorscols; c++) {
+                    res2->set(r,c,eigenVectors.coeff(r, c));
+    }
+    }
+    }
+};
```

Comment:
Likewise the output does not need to be copied. There are DenseMatrix constructors taking chucks of memory and I'm sure there's a way to get the data pointer out of an Eigen matrix.

Reply:
Thinking twice about it it'd be better to allocate the DenseMatrix and pass its data pointer to an Eigen matrix. Otherwise your memory for the output would be held by a local variable that gets deallocated at the end of your function.